extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/dataconnection/MongoDataConnection.java

```diff
@@ -185,9 +195,13 @@ private static void addResources(List<DataConnectionResource> resources, MongoDa
      */
     @Override
     public void destroy() {
-        if (mongoClient != null) {
-            ((CloseableMongoClient) mongoClient).unwrap().close();
-            mongoClient = null;
+        ConcurrentMemoizingSupplier<MongoClient> supplier = mongoClientSup;
+        if (supplier != null) {
+            MongoClient mongoClient = mongoClientSup.remembered();
+            if (mongoClient != null) {
+                ((CloseableMongoClient) mongoClient).unwrap().close();
+            }
+            mongoClientSup = null;
```

Comment:
If you nullify the `mongoClientSup` *after* closing the client, you have supplier returning closed client.